### PR TITLE
Arena match result

### PR DESCRIPTION
**Pull request by AminArria**
- [x] Schema to hold general match results per player
- [x] Endpoint to save a single player match result
  - [x] Modify endpoint to accept a list of results per player (to avoid needing a request per result per player)
- [x] Send match results on arena game end 
  - [x] Identify regular players from bot players when processing results to avoid sending bot results
- [x] Rebase PR after #532 is merged
- [ ] (optional) Setup `GATEWAY_URL` in servers, if not present nothing bad will happen only `game_client` login won't work and match result reporting won't work, nothing game breaking
- [x] Add login flow in `game_client` 

### Testing
To test this PR you can play a game through the browser client (`game_client`) and then check the DB with `GameBackend.Repo.all(GameBackend.Matches.ArenaMatchResult)`, it should show you only your single match result (bots are not recorded)

**NOTE** To test this with login you will need to setup `GOOGLE_CLIENT_ID` and `GOOGLE_CLIENT_SECRET` (they are in our 1password vault)

Another way to test this pr is creating a new google user with:
`GameBackend.Users.find_or_create_google_user_by_email("a")`

Copy that google user id and paste it here:

`http://localhost:3000/board/CREATED_GOOGLE_ID/uma/jorge/join`

Play the match and when you finish you should have a match result registry in your db

### CODE REVIEW

File: apps/game_backend/priv/repo/migrations/20240425171037_add_arena_match_results.exs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+defmodule GameBackend.Repo.Migrations.AddArenaMatchResults do
+  use Ecto.Migration
+
+  def change do
+    create table(:arena_match_results) do
+      add :user_id, references(:google_users, on_delete: :delete_all), null: false
```
Comment (AminArria) [inline apps/game_backend/priv/repo/migrations/20240425171037_add_arena_match_results.exs:6]: Index?

File: apps/arena/lib/arena/game_updater.ex [hunk 2]
```diff
@@ -485,6 +488,38 @@ defmodule Arena.GameUpdater do
     |> Map.put(:aditional_info, entity |> Entities.maybe_add_custom_info())
   end
 
+  defp report_game_results(state, winner_id) do
+    results =
+      Enum.filter(state.game_state.client_to_player_map, fn {client_id, _player_id} ->
+        Enum.member?(state.clients, client_id)
+      end)
```
Comment (agustinesco) [inline apps/arena/lib/arena/game_updater.ex:495]: you can also use `Map.take(state.game_state.client_to_player_map, state.clients)`

File: apps/arena/lib/arena/game_updater.ex [hunk 3]
```diff
@@ -485,6 +488,38 @@ defmodule Arena.GameUpdater do
...
+        }
+      end)
+
+    payload = Jason.encode!(%{match_id: Ecto.UUID.generate(), results: results})
+
+    ## TODO: we should be doing this in a better way, both the url and the actual request
+    ## maybe a separate GenServer that gets the results and tries to send them to the server?
+    ## This way if it fails we can retry or something
+    spawn(fn ->
+      gateway_url = System.get_env("GATEWAY_URL") || "http://localhost:4001"
```
Comment (agustinesco) [inline apps/arena/lib/arena/game_updater.ex:516]: should this be in the configuration module?

File: apps/arena/lib/arena/game_updater.ex [hunk 4]
```diff
@@ -485,6 +488,38 @@ defmodule Arena.GameUpdater do
...
+
+      Finch.build(:post, "#{gateway_url}/arena/match", [{"content-type", "application/json"}], payload)
+      |> Finch.request(Arena.Finch)
```
Comment (agustinesco) [inline apps/arena/lib/arena/game_updater.ex:519]: remember that Finch.request is sync and could cause the game_updater to timeout if the game_backend app isn't reachable, you can replace this with `async_request` or spawn a new process to handle this request, pretty similar to what `game_launcher` is doing

File: apps/game_backend/lib/game_backend/matches.ex [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  def create_arena_match_results(match_id, results) do
+    transaction = Multi.new()
+
+    {transaction, _} =
+      Enum.reduce(results, {transaction, 1}, fn result, {transaction_acc, idx} ->
+        attrs = Map.put(result, "match_id", match_id)
+        changeset = ArenaMatchResult.changeset(%ArenaMatchResult{}, attrs)
+        transaction_acc = Multi.insert(transaction_acc, {:insert, idx}, changeset)
+        {transaction_acc, idx + 1}
+      end)
```
Comment (agustinesco) [inline apps/game_backend/lib/game_backend/matches.ex:18]: i think you can omit this indexing by using the client id in the transaction key:

```
 Enum.reduce(results, Multi.new(), fn result, transaction_acc ->
        attrs = Map.put(result, "match_id", match_id)
        changeset = ArenaMatchResult.changeset(%ArenaMatchResult{}, attrs)
        Multi.insert(transaction_acc, {:insert, attrs.user_id}, changeset)
      end)
```

Comment (agustinesco) [reply apps/game_backend/lib/game_backend/matches.ex:18]: also, if we're going to include the match id in the registry why send them in a separate field? maybe we could include the match_id when we create the attrs map in arena an match the whole map in `match_results_controller`

File: apps/arena/lib/arena/game_updater.ex (hunk 4 above)
Comment (AminArria) [reply apps/arena/lib/arena/game_updater.ex:519]: This is inside a `spawn/1` so no issues with that

Comment (agustinesco) [reply apps/arena/lib/arena/game_updater.ex:519]: oh right, i didn't saw it 😅

File: apps/game_backend/lib/game_backend/matches.ex (hunk 5 above)
Comment (AminArria) [reply apps/game_backend/lib/game_backend/matches.ex]: Changed [622cd6d](https://github.com/lambdaclass/mirra_backend/pull/529/commits/622cd6d101e20b92d86fde749ce12600b3598d30)

File: apps/arena/lib/arena/game_updater.ex (hunk 3 above)
Comment (AminArria) [reply apps/arena/lib/arena/game_updater.ex]: Yes, changed [622cd6d](https://github.com/lambdaclass/mirra_backend/pull/529/commits/622cd6d101e20b92d86fde749ce12600b3598d30)

File: apps/arena/lib/arena/game_updater.ex (hunk 2 above)
Comment (AminArria) [reply apps/arena/lib/arena/game_updater.ex]: Nice tip! [622cd6d](https://github.com/lambdaclass/mirra_backend/pull/529/commits/622cd6d101e20b92d86fde749ce12600b3598d30)

File: apps/game_client/lib/game_client_web/controllers/page_controller.ex [hunk 6]
```diff
@@ -2,10 +2,12 @@ defmodule GameClientWeb.PageController do
...
-    render(conn, :home)
+    user_id = get_session(conn, :user_id)
+    render(conn, :home, user_id: user_id)
   end
 
-  def select_character(conn, %{"character" => character, "game_mode" => game_mode}) do
-    redirect(conn, to: ~p"/board/#{Ecto.UUID.generate()}/#{character}/player_name/#{game_mode}")
+  def select_character(conn, %{"character" => character, "game_mode" => game_mode} = params) do
+    user_id = params["user_id"] || Ecto.UUID.generate()
+    redirect(conn, to: ~p"/board/#{user_id}/#{character}/player_name/#{game_mode}")
```
Comment (agustinesco) [inline apps/game_client/lib/game_client_web/controllers/page_controller.ex:11]: This is not working if you don't have user_id

**Review by agustinesco — APPROVED** [review]
LGTM!

**Comment by agustinesco** [comment]
I've just noticed that when you play a match in the unity client the Gateway application crashes, should we handle this?

```
[debug] Converted error Ecto.ChangeError to 500 response
[error] #PID<0.5915.0> running Gateway.Endpoint (connection #PID<0.5914.0>, stream id 1) terminated
Server: localhost:4001 (http)
Request: POST /arena/match
** (exit) an exception was raised:
    ** (Ecto.ChangeError) value `"1758441033"` for `GameBackend.Matches.ArenaMatchResult.user_id` in `insert` does not match type :binary_id
        (ecto 3.11.1) lib/ecto/repo/schema.ex:1049: Ecto.Repo.Schema.dump_field!/6
```

**Review by agustinesco — APPROVED** [review]
LGTM!

**Review by ncontinanza — APPROVED** [review]

**Pull request merged**